test/e2e/storage/api/testsuite.go

```diff
@@ -0,0 +1,130 @@
+			suite.skipUnsupportedTests(driver, pattern)
+		})
+		// actually define the tests
+		// at this step the testsuite should not worry about if the pattern and driver
+		// does not fit for the whole testsuite. But driver&pattern check
+		// might still needed for specific independent test cases.
+		suite.defineTests(driver, pattern)
+	})
+}
+
+// DefineTestSuites defines tests for all testpatterns and all testSuites for a driver
+func DefineTestSuites(driver TestDriver, tsInits []func() TestSuite) {
+	for _, testSuiteInit := range tsInits {
+		suite := testSuiteInit()
+		for _, pattern := range suite.GetTestSuiteInfo().TestPatterns {
+			RegisterTests(suite, driver, pattern)
+		}
+	}
+}
+
+// TestSuiteInfo represents a set of parameters for TestSuite
+type TestSuiteInfo struct {
+	Name               string              // name of the TestSuite
+	FeatureTag         string              // featureTag for the TestSuite
+	TestPatterns       []TestPattern       // Slice of TestPattern for the TestSuite
+	SupportedSizeRange e2evolume.SizeRange // Size range supported by the test suite
+}
+
+// SkipInvalidDriverPatternCombination will skip tests if the combination of driver, and testpattern
+// is not compatible to be tested. This function will be called in the TestSuiteHandler to make
+// sure all the testsuites we defined are valid.
+//
+// Whether it needs to be skipped is checked by following steps:
+// 0. Check with driver SkipUnsupportedTest
+// 1. Check if volType is supported by driver from its interface
+// 2. Check if fsType is supported
+//
+// Test suites can also skip tests inside their own skipUnsupportedTests function or in
+// individual tests.
+func SkipInvalidDriverPatternCombination(driver TestDriver, pattern TestPattern) {
+	dInfo := driver.GetDriverInfo()
+	var isSupported bool
+
+	// 0. Check with driver specific logic
+	driver.SkipUnsupportedTest(pattern)
+
+	// 1. Check if Whether volType is supported by driver from its interface
+	switch pattern.VolType {
+	case InlineVolume:
+		_, isSupported = driver.(InlineVolumeTestDriver)
+	case PreprovisionedPV:
+		_, isSupported = driver.(PreprovisionedPVTestDriver)
+	case DynamicPV, GenericEphemeralVolume:
+		_, isSupported = driver.(DynamicPVTestDriver)
+	case CSIInlineVolume:
+		_, isSupported = driver.(EphemeralTestDriver)
+	default:
+		isSupported = false
+	}
+
+	if !isSupported {
+		e2eskipper.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolType)
+	}
+
+	// 2. Check if fsType is supported
+	if !dInfo.SupportedFsType.Has(pattern.FsType) {
+		e2eskipper.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.FsType)
+	}
+	if pattern.FsType == "xfs" && framework.NodeOSDistroIs("windows") {
+		e2eskipper.Skipf("Distro doesn't support xfs -- skipping")
+	}
+	if pattern.FsType == "ntfs" && !framework.NodeOSDistroIs("windows") {
+		e2eskipper.Skipf("Distro %s doesn't support ntfs -- skipping", framework.TestContext.NodeOSDistro)
```

Comment:
I am thinking after I separate in-tree windows gce pd driver, we don't need this check. A driver should specify what fsType it support.

Reply:
We need to check other intree cloud drivers too. It's also possible aws and azure are running these e2es outside of Kubernetes. cc @andyzhangx @wongma7 @jsafrane

Reply:
I think using fstype and OS to check is not ideal considering in the future, same fstype might be supported by both linux and windows for some drivers.

Not part of your change. I am proposing to have supportedOSes in driver configuration. It explicitly specify what OSes it can support and the testsuits will check based on framework.NodeDistrol value. The default value can be set to "linux".